activerecord/lib/active_record/associations/has_many_association.rb

```diff
@@ -84,7 +84,10 @@ def count_records
           # If there's nothing in the database and @target has no new records
           # we are certain the current target is an empty array. This is a
           # documented side-effect of the method that may avoid an extra SELECT.
-          loaded! if count == 0
+          if count == 0 && target.select(&:new_record?).empty?
+            self.target = []
+            loaded!
+          end
```

Comment:
Does everything still work as expected if we apply this optimization even with `new_record?` records?  i.e.:

```suggestion
          if count == 0
            target.select!(&:new_record?)
            loaded!
          end
```